src/librustdoc/clean/types.rs

```diff
@@ -388,6 +388,64 @@ impl Item {
         self.def_id.as_def_id().map(|did| tcx.get_attrs(did).inner_docs()).unwrap_or(false)
     }
 
+    crate fn visibility(&self, tcx: TyCtxt<'_>) -> Visibility {
+        let mut def_id = match self.def_id {
+            // Anything but DefId *shouldn't* matter, but return a reasonable value anyway.
+            ItemId::Auto { .. } | ItemId::Blanket { .. } => return Visibility::Inherited,
+            ItemId::Primitive(..) => return Visibility::Public,
+            ItemId::DefId(did) => did,
+        };
+        match *self.kind {
+            // Variant fields inherit their enum's visibility.
+            StructFieldItem(..) if is_field_vis_inherited(tcx, def_id) => {
+                return Visibility::Inherited;
+            }
+            // Variants always inherit visibility
+            VariantItem(..) => return Visibility::Inherited,
+            // Return the visibility of `extern crate` statement, not the crate itself (which will always be public).
+            ExternCrateItem { .. } => {
+                use rustc_session::cstore::ExternCrateSource;
+
+                // `extern crate self` will ICE if it's called with `extern_crate`;
+                // special-case it.
+                if !def_id.is_local() {
```

Comment:
This is a bug, it needs to get the DefId of the `extern crate self` statement somehow.

Reply:
apparently you can have multiple `extern crate x` statements ?? https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/How.20can.20I.20get.20the.20DefId.20of.20an.20.60extern.20crate.20self.60.20statement.3F
going to change this to a field on `ExternCrateItem` for now, it seems less fragile